pkg/services/annotations/annotationsimpl/annotations.go

```diff
@@ -72,12 +72,42 @@ func (r *RepositoryImpl) Update(ctx context.Context, item *annotations.Item) err
 }
 
 func (r *RepositoryImpl) Find(ctx context.Context, query *annotations.ItemQuery) ([]*annotations.ItemDTO, error) {
-	resources, err := r.authZ.Authorize(ctx, query.OrgID, query)
-	if err != nil {
-		return make([]*annotations.ItemDTO, 0), err
+	// Search by tags is expensive, so check without access control first
+	if len(query.Tags) > 0 && query.DashboardID == 0 && query.DashboardUID == "" {
+		// Return early if no annotations found, it's not necessary to perform expensive access control filtering
+		res, err := r.reader.Get(ctx, query, &accesscontrol.AccessResources{
+			SkipAccessControlFilter: true,
+		})
+		if err != nil || len(res) == 0 {
+			return []*annotations.ItemDTO{}, err
+		}
 	}
 
-	return r.reader.Get(ctx, query, resources)
+	results := make([]*annotations.ItemDTO, 0, query.Limit)
+	query.Page = 1
+
+	// Iterate over available annotations util query limit is reached
+	// or all available dashboards are checked
+	for len(results) < int(query.Limit) {
+		resources, err := r.authZ.Authorize(ctx, query)
+		if err != nil {
+			return nil, err
+		}
+
+		res, err := r.reader.Get(ctx, query, resources)
```

Comment:
Aaron raised a good question about us passing a page in to the annotation reader store. I think it should not change anything, but it might be a bit tricky, as there are a bunch of annotation stores and I'm not familiar with their behaviour. Might be safer to pass in an unmodified query without a page specified here.

Reply:
That's a good point, we don't need to paginate Get() query, only authorization one.

Reply:
Ah, double-checked. We didn't use Page param in store. I added it in this PR and it's only used inside `Authorize()` method. So should be safe.